File: csharp/Chapter08/RunAddClient.cs
using System;
using com.db4o;

namespace com.db4o.dg2db4o.chapter8
{
    class RunAddClient
    {
        private ObjectContainer aClient;
        private bool stop = false;

        public void Run()
        {
            lock (this)
            {
                Console.WriteLine("Starting add client...");
                aClient = Db4o.OpenClient("localhost", 8732, "user1", "password");
                while (!stop)
                {
                    Console.WriteLine("ADDCLIENT: Please enter a name: ");
                    String name = Console.ReadLine();
                    Console.WriteLine("ADDCLIENT: Please enter an age: ");
                    int age = Convert.ToInt32(Console.ReadLine());
                    aClient.Set(new Person(name, age));
                    aClient.Commit();
                }
                aClient.Close();
            }
        }

    }
}

File: java/src/com/db4o/dg2db4o/chapter8/RunAddClient.java

package com.db4o.dg2db4o.chapter8;

import java.util.Scanner;
import com.db4o.*;
import com.db4o.messaging.*;

public class RunAddClient implements Runnable {
    private boolean stop = false; // Use this flag to quit
    ObjectContainer aClient = null;
    Scanner sc = new Scanner(System.in);
    
    // Can run as standalone process, or use Main.java to run as thread
    public static void main(String[] args) {
        RunAddClient runner = new RunAddClient();
        runner.run();
    }
    
    public void run() {
        synchronized(this){
            try {
                aClient = Db4o.openClient("localhost", 8732, "user1", "password");
                while (!stop) {
                    System.out.print("\nADDCLIENT:Enter the developer's name and age (e.g. 'Tom 44'): ");
                    String name = sc.next();
                    if(name.equals("stop"))
                        stop = true;
                    else {
                        aClient.set(new Person(name, sc.nextInt()));
                        aClient.commit();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                MessageSender messageSender = aClient.ext().configure()
                .getMessageSender();
                messageSender.send(new StopServer("ADDCLIENT says stop!"));
                aClient.close();
            }
        }
    }
}

